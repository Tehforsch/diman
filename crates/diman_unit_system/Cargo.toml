[package]
name = "diman_unit_system"
version = "0.2.0"
edition = "2021"
authors = [
    "Toni Peter <mail@tonipeter.de>",
]

[features]
glam = []
glam-vec2 = []
glam-dvec2 = []
glam-vec3 = []
glam-dvec3 = []
f32 = []
f64 = []
mpi = []
serde = []
rand = []
hdf5 = []
gen-vec-names = [] 

[dependencies]
derive_verify = { path = "../../crates/derive_verify" }
syn = { version = "1.0", features = ["full", "extra-traits"] }
quote = "1.0"
proc-macro2 = "1.0"

[lib]
proc-macro = true
