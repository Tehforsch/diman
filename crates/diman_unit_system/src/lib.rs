#![feature(proc_macro_diagnostic)]

mod codegen;
mod dimension_math;
mod parse;
mod resolve;
mod types;

use proc_macro2::TokenStream;
use syn::*;

#[proc_macro]
pub fn unit_system(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let defs = parse_macro_input!(item as types::UnresolvedTemplates);
    let expanded = defs.expand_templates();
    let resolved = expanded.resolve();
    let impls: TokenStream = resolved.code_gen();
    impls.into()
}
