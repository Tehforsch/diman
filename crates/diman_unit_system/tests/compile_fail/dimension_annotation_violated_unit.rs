#![allow(incomplete_features)]
#![feature(generic_const_exprs, adt_const_params)]

use diman_unit_system::unit_system_internal;
unit_system_internal!(
    quantity_type Quantity;
    dimension_type Dimension;
    dimension Mass;
    dimension Length;
    #[base(Mass)]
    #[symbol(kg)]
    unit kilograms: Mass;
    unit foo: Length = kilograms * kilograms;
);

fn main() {
}
