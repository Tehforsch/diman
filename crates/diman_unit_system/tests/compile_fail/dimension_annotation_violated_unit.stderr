error: Dimension mismatch in expression.
  --> tests/compile_fail/dimension_annotation_violated_unit.rs:13:15
   |
13 |     unit foo: Length = kilograms * kilograms;
   |               ^^^^^^
   |
   = help: The annotation on the left-hand side has dimensions length^1 mass^0
   = help: but the expression on the right-hand side has dimensions length^0 mass^2
